docs/edgeql/path_resolution.rst

```diff
@@ -0,0 +1,151 @@
+.. _ref_eql_path_resolution:
+
+===============
+Path resolution
+===============
+
+Element-wise operations with multiple arguments in EdgeDB are generally applied
+to the :ref:`cartesian product <ref_reference_cardinality_cartesian>` of all
+the input sets.
+
+.. code-block:: edgeql-repl
+
+    db> select {'aaa', 'bbb'} ++ {'ccc', 'ddd'};
+    {'aaaccc', 'aaaddd', 'bbbccc', 'bbbddd'}
+
+In some cases, this works out fine, but in others it doesn't make sense. Take
+this example:
+
+.. code-block:: edgeql
+
+    select User.first_name ++ ' ' ++ User.last_name;
+
+Should the result of this query be every ``first_name`` in your ``User``
+objects concatenated with every ``last_name``? That's probably not what you
+want.
+
+This is why, in cases where multiple element-wise arguments share a common path
+(``User.`` in this example), EdgeDB factors out the common path rather than
+using cartesian multiplication.
+
+.. code-block:: edgeql-repl
+
+    db> select User.first_name ++ ' ' ++ User.last_name;
+    {'Mina Murray', 'Jonathan Harker', 'Lucy Westenra', 'John Seward'}
+
+We assume this is what you want, but if your goal is to get the cartesian
+product, you can accomplish it one of two ways. You could use
+:eql:op:`detached`.
+
+.. code-block:: edgeql-repl
+
+    edgedb> select User.first_name ++ ' ' ++ detached User.last_name;
+    {
+      'Mina Murray',
+      'Mina Harker',
+      'Mina Westenra',
+      'Mina Seward',
+      'Jonathan Murray',
+      'Jonathan Harker',
+      'Jonathan Westenra',
+      'Jonathan Seward',
+      'Lucy Murray',
+      'Lucy Harker',
+      'Lucy Westenra',
+      'Lucy Seward',
+      'John Murray',
+      'John Harker',
+      'John Westenra',
+      'John Seward',
+    }
+
+Or you could use :ref:`with <ref_eql_with>` to attach a different symbol to
+your set of ``User`` objects.
+
+.. code-block:: edgeql-repl
+
+    edgedb> with U := (select User)
+    ....... select U.first_name ++ ' ' ++ User.last_name;
+    {
+      'Mina Murray',
+      'Mina Harker',
+      'Mina Westenra',
+      'Mina Seward',
+      'Jonathan Murray',
+      'Jonathan Harker',
+      'Jonathan Westenra',
+      'Jonathan Seward',
+      'Lucy Murray',
+      'Lucy Harker',
+      'Lucy Westenra',
+      'Lucy Seward',
+      'John Murray',
+      'John Harker',
+      'John Westenra',
+      'John Seward',
+    }
+
+The reason ``with`` works here even though the alias ``U`` refers to the exact
+same set is that we only assume you want the path factored in this way when you
+use the same *symbol* to refer to a set. This means operations with
+``User.first_name`` and ``User.last_name`` do get the common path factored
+while ``U.first_name`` and ``User.last_name`` do not and are resolved with
+cartesian multiplication.
```

Comment:
I am not sure but it seems that there are better explanation, that the clauses of with and the body of with live in different scopes (same reason as for the next section). Maybe Sully can confirm?

Reply:
The technical reason is that `U` in `U.first_name` is really a binder that has not been instantiated. Putting it on a par with `User.last_name` seems strange.

Reply:
Any ideas for a better way to explain this @vpetrovykh or @msullivan? This was my takeaway from my conversation with Victor. I may have misunderstood.

Reply:
Both of these reasons are correct and are part of what is going on:
 * `U.first_name` and `User.last_name` don't have anything factored out because they don't have a common prefix
 * The `User` in the `with` binding and the `User` in the `select` body don't get factored because they are in separate subqueries

Reply:
Makes sense. I will update it to reflect that. That will give me a natural segue into scopes too. Thanks @zchenb and @msullivan!